File: 2.2 Linear DS/2D Array Manipulation/p10855.py
# Competitive Programming 3
# Problem 10855

while True:
    N, n = [int(x) for x in input().split()]
    if N == 0 and n == 0: break

    big = ['0'*N for _ in range(N)]
    small = ['0'*n for _ in range(n)]
    
    for i in range(N): big[i] = input()
    for i in range(n): small[i] = input()
        
    
    count = {0:0, 90:0, 180:0, 270:0}
    
    for i in range(N-n+1):
        for j in range(N-n+1):
            # correct orientation
            match = True
            for x in range(n):
                for y in range(n):
                    
                    if big[i+x][j+y] != small[x][y]:
                        match = False
                        break
                if not match: break
            if match: count[0] += 1
            
            # 90 degrees clockwise
            match = True
            for x in range(n):
                for y in range(n):
                    if big[i+x][j+y] != small[n-1-y][x]:
                        match = False
                        break
                if not match: break
            if match: count[90] += 1

            # 180 degrees clockwise
            match = True
            for x in range(n):
                for y in range(n):
                    if big[i+x][j+y] != small[n-1-y][n-1-x]:
                        match = False
                        break
                if not match: break
            if match: count[180] += 1

            # 270 degrees clockwise
            match = True
            for x in range(n):
                for y in range(n):
                    if big[i+x][j+y] != small[y][n-1-x]:
                        match = False
                        break
                if not match: break
            if match: count[270] += 1

    print(*[count[i] for i in range(0, 360, 90)])
                        
                    
                

File: 2.2 Linear DS/2D Array Manipulation/p10855.cpp
// Competitive Programming 3
// Problem 10855

#include <iostream>
using namespace std;

int main(){
    while (1){
        int N, n;
        char c;
        cin >> N >> n;
        if (N == 0 && n == 0) break;

        int big[N][N], small[n][n];
        for (int i = 0; i < N; i++){
            for (int j = 0; j < N; j++){
                cin >> c;
                big[i][j] = (int) c;
            }
        }
        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                cin >> c;
                small[i][j] = (int) c;
            }
        }

        int count[] = {0,0,0,0};

        // correct orientation
        for (int i = 0; i < N-n+1; i++){
            for (int j = 0; j < N-n+1; j++){
                bool match = true;
                for (int x = 0; x < n; x++){
                    for (int y = 0; y < n; y++){
                        if (big[i+x][j+y] != small[x][y]){
                            match = false;
                            break;
                        }
                    }
                    if (!match) break;
                }
                if (match) count[0] += 1;
            }
        }

        // rotated 90 degrees
        for (int i = 0; i < N-n+1; i++){
            for (int j = 0; j < N-n+1; j++){
                bool match = true;
                for (int x = 0; x < n; x++){
                    for (int y = 0; y < n; y++){
                        if (big[i+x][j+y] != small[n-1-y][x]){
                            match = false;
                            break;
                        }
                    }
                    if (!match) break;
                }
                if (match) count[1] += 1;
            }
        }

        // rotated 180 degrees
        for (int i = 0; i < N-n+1; i++){
            for (int j = 0; j < N-n+1; j++){
                bool match = true;
                for (int x = 0; x < n; x++){
                    for (int y = 0; y < n; y++){
                        if (big[i+x][j+y] != small[n-1-x][n-1-y]){
                            match = false;
                            break;
                        }
                    }
                    if (!match) break;
                }
                if (match) count[2] += 1;
            }
        }

        // rotated 270 degrees
        for (int i = 0; i < N-n+1; i++){
            for (int j = 0; j < N-n+1; j++){
                bool match = true;
                for (int x = 0; x < n; x++){
                    for (int y = 0; y < n; y++){
                        if (big[i+x][j+y] != small[y][n-1-x]){
                            match = false;
                            break;
                        }
                    }
                    if (!match) break;
                }
                if (match) count[3] += 1;
            }
        }

        cout << count[0] << ' ' << count[1] << ' ';
        cout << count[2] << ' ' << count[3] << endl;


    }
    return 0;
}
